Source/SwiftLintBuiltInRules/Rules/Lint/UnneededSynthesizedInitializerRule.swift

```diff
@@ -0,0 +1,286 @@
+    ) -> Bool {
+        guard initializerParameters.isNotEmpty else { return noParameterInitializer(storedProperties) }
+        guard initializerParameters.count == storedProperties.count else { return false }
+
+        for (idx, parameter) in initializerParameters.enumerated() {
+            guard parameter.secondName == nil else { return false }
+
+            let property = storedProperties[idx]
+            let propertyId = property.firstIdentifier
+            let propertyType = property.bindings.first?.typeAnnotation?.type
+
+            // Ensure that parameters that correspond to properties declared using 'var' have a default
+            // argument that is identical to the property's default value. Otherwise, a default argument
+            // doesn't match the memberwise initializer.
+            let isVarDecl = property.bindingKeyword.tokenKind == .keyword(.var)
+            if isVarDecl, let initializer = property.bindings.first?.initializer {
+                guard let defaultArg = parameter.defaultArgument else { return false }
+                guard initializer.value.description == defaultArg.value.description else { return false }
+            } else if parameter.defaultArgument != nil {
+                return false
+            }
+
+            if
+                propertyId.identifier.text != parameter.firstName.text
+                    || (propertyType != nil && propertyType?.description.trimmingCharacters(in: .whitespaces) !=
+                    parameter.type.description.trimmingCharacters(in: .whitespacesAndNewlines))
+            {
+                return false
+            }
+        }
+        return true
+    }
+
+    // Does the body initialize all, and only, the stored properties for the struct?
+    private func initializerBody( // swiftlint:disable:this cyclomatic_complexity
+        _ initializerBody: CodeBlockSyntax?,
+        matches storedProperties: [VariableDeclSyntax]
+    ) -> Bool {
+        guard let initializerBody else { return false }
+        guard storedProperties.count == initializerBody.statements.count else { return false }
+
+        var statements: [String] = []
+        for statement in initializerBody.statements {
+            guard let exp = statement.item.as(SequenceExprSyntax.self) else { return false }
+            var leftName = ""
+            var rightName = ""
+
+            for element in exp.elements {
+                switch Syntax(element).as(SyntaxEnum.self) {
+                case .memberAccessExpr(let element):
+                    guard let base = element.base,
+                          base.description.trimmingCharacters(in: .whitespacesAndNewlines) == "self"
+                    else {
+                        return false
+                    }
+                    leftName = element.name.text
+                case .assignmentExpr(let element):
+                    guard element.assignToken.tokenKind == .equal else { return false }
```

Comment:
So 

```
case .assignmentExpr(let element) where element.assignToken.tokenKind != .equal:
    return false
case .identifierExpr(let element):
    rightName = element.identifier.text
default:
    return false
}
```

does not work the same, because any `.assignmentExpr` that does not match the `where` clause will end up returning `false` in the `default` case, which is wrong.

To work, we'd need something like

```
case .assignmentExpr(let element) where element.assignToken.tokenKind != .equal:
    return false
case .assignmentExpr(_):
    break
case .identifierExpr(let element):
    rightName = element.identifier.text
default:
    return false
}
```

more than happy to change it to that, but at that point is it actually better/clearer than the original (ported) code?

Reply:
The second (and implemented) one looks clearer to me. Thanks!